Protocol for extending the variables pane

dfalbel opened the pull request:
This PR  is a proposal of a mechanism that allows package authors to implement custom behavior for objects in the variables pane. 

A description of how the protocol works is available here: https://github.com/posit-dev/ark/blob/daf9348affc0f1b190928a2fb0cad5bfadaf263c/doc/variables-pane-extending.md

See also #561 for the custom inspect behavior.

## Motivation

The main motivation is to allow package authors to improve the display of objects for which the internal R data structure does not show useful information for users. For example, reticulate R obejcts are environments containing a flag and an external pointer (`x <- reticulate::py_eval("1", convert = FALSE)`):

![image](https://github.com/user-attachments/assets/309e8771-f707-415f-8f9b-cbd0d3947429)

Or torch tensors (`x = torch::torch_randn(10, 10)`) :

![image](https://github.com/user-attachments/assets/093348b5-21fc-4c53-a7c1-5109777a339e)

Or xml2 documents (`x <- xml2::read_xml("<foo> <bar> text <baz/> </bar> </foo>")`):

![image](https://github.com/user-attachments/assets/d3230a84-ae4a-480f-971b-b729bcb847be)

## How it's implemented

Package authors can implement a set of S3 like methods, each customizing a specific part of the UI.
Methods are discovered by ark when packages are loaded and lazily stored in an environment. Package authors don't need to export these methods in their packages. See [here](https://github.com/posit-dev/ark/blob/daf9348affc0f1b190928a2fb0cad5bfadaf263c/doc/variables-pane-extending.md) for the description of methods.

When responding to events from the variables pane, we try to apply the custom methods and if the method works correctly we use it's value, otherwise we try to continue using the default implementation.

Methods are expllicitly implemented for Ark, so it's expected that package authors will take extra care to make they work correctly with Positron.

## Comparison to RStudio

RStudio uses the `utils::str` S3 in a few situations to obtain the values description, see usages of ValueFromStr in eg: https://github.com/rstudio/rstudio/blob/2175496934a54e21979461a1c67eb13ad5048cf2/src/cpp/session/modules/SessionEnvironment.R#L122

There are many guardrails though, as `str` methods are commonly implemented and package authors might not be aware of such bugs when used from RStudio.

dfalbel commented:
recheck

dfalbel commented:
I have read the CLA Document and I hereby sign the CLA

dfalbel commented:
recheck

t-kalinowski commented:
I have read the CLA Document and I hereby sign the CLA

lionel- commented:
That seems like a good idea!

Before reviewing the implementation, I'd just like to discuss a potential simplification of the extension mechanism. What if we had a single generic `ps_variable_proxy()` that would be in charge of converting an object to a base type representation (data frame, vector, list, environment) that Ark would be able to display in the usual way. This is consistent with how genericity works in the tidyverse via vctrs proxies.

The display type could be set to the first element of the class vector, but if that's too inconvenient we could have an additional method for it as proposed here.

Laziness for deeply/infinitely nested objects:

- Could be supported via environments. The only downside is that environments have the restriction of uniquely named elements, but that seems reasonable for object internals since that's consistent with struct-like types?

- Or we could allow list-proxies containing classed objects that themselves have a proxy method.

What do you think?

lionel- reviewed (commented):
That seems like a good idea!

Before reviewing the implementation, I'd just like to discuss a potential simplification of the extension mechanism. What if we had a single generic `ps_variable_proxy()` that would be in charge of converting an object to a base type representation (data frame, vector, list, environment) that Ark would be able to display in the usual way. This is consistent with how genericity works in the tidyverse via vctrs proxies.

The display type could be set to the first element of the class vector, but if that's too inconvenient we could have an additional method for it as proposed here.

Laziness for deeply/infinitely nested objects:

- Could be supported via environments. The only downside is that environments have the restriction of uniquely named elements, but that seems reasonable for object internals since that's consistent with struct-like types?

- Or we could allow list-proxies containing classed objects that themselves have a proxy method.

What do you think?

dfalbel commented:
I like the idea of a single ps_variable_proxy.

I think I'd still want to customize the display value though, eg for a torch_tensor, I think the display value should be something like `Float [1:10, 1:10]`, I think the closest proxy would be a length 1 character vector with that text, which would still be displayed with quotes around it:`"Float [1:10, 1:10]"`. 

I'm fine with `display type = class(x)[1]`, we could remove `display_type`.

So the proxy would replace `has_children()`, `get_children()` and `get_child_at()`.

My main concern with a single proxy implementation is performance. The way the variables pane is currenty implemented, we don't keep any state after the representations are sent to the front-end and we only get back a vector of access keys from the front-end when we want to expand a node.

Considering we implement the second option where variable_proxy() returns a list-proxy containing classed objects implementing `variable_proxy`. So suppose that A is an object that implements the `variable_proxy()` and it's children are also objects that also implement `variable_proxy()` (For instance, A could be a reticulate object representing a Python dictionary and B a dictionary contained in A).

When we click on B, so we get it's children, we'll receive a request from the front-end containing [A, B].
To compute the children of B, we'll need to:

1. `ps_variable_proxy(A)`
2. find `B` there and then
3. `ps.variable_proxy(B)`

```
Root
└── A
    ├── B
    │   ├── B1
    │   └── B2
    └── C
```

If `A` is large, building it's `variable_proxy()` can be time consuming and we'll need to rebuild it for every. neste child one expands.  

The main reason for having two separate `get_children()` and `get_child()` is that there might be an efficient way of finding `B` in `A` without building the entire list of possible children.

IMO this will be a rare situation though, except for reticulate, most use cases I can think would only want to show a nice display value. So maybe it's fine to let package authors guarantee that variable_proxy is fast enough, even if they need to implement some kind of caching, etc. Or maybe, we want to keep some state in the variables pane so subsequent requests don't need to recompute the children.

Let me know what you think!

lionel- commented:
hmm if performance is an issue I think we should be able to store the proxies in our `current_bindings`? These would only be updated if the objects change across top-level calls.

lionel- commented:
In my mind the variable pane should never contain GBs of data and any materialised data should be truncated - to support large data we'd need a similar approach to the data viewer and then the proxy becomes a slice proxy.

I see your points regarding flexibility of data types in the context of interop. I think this aligns with Daniel's request to keep a display value extension point?

t-kalinowski commented:
> I think this aligns with Daniel's request to keep a display value extension point?

I'm probably misunderstanding, but I don't think it does. A code example might help. 

With the proxy approach, what method(s) would reticulate implement to display a NumPy array?

dfalbel commented:
I think for a numpy array, reticulate would only implement `display_value` and a proxy method that returns anything that's scalar atomic, so the arrow `\/` to list children doesn't appear for it in the variables pane (equivalent to current `has_children() = FALSE`).

Storing proxies in `current_bindings` is probably possible, but the implementation will be quite tricky, because an object implementing `variable_proxy()` could be any level nested into an object that is pointed by one of the `current_bindings`.
I can take a look at what that could look like.

lionel- commented:
@dfalbel Feel free to take a look but I think you've both convinced me that the proxy approach would not be a simplification.

dfalbel commented:
Ok, I didn't get to it. I set up an example package here: https://github.com/dfalbel/testVariablesPaneExtension/tree/main/R with an example that works for R6 classes and some examples of what I'd like to make it work for torch.

lionel- reviewed (approved):
This is a really nice PR!
Approved but I think @DavisVaughan should also take a look.

I believe this is the very first API exposed from Ark to R packages (if we exclude rstudioapi support) so we should discuss what that should look like. I like the setup you've created to register methods but I wonder if:

```r
.ps.register_ark_method("ark_variable_display_value", ...)
```

should have the opposite namespacing:

```r
.ark.register_method("positron_variable_display_value", ...)
```

- The variable pane is currently a positron-only feature so should be namespaced in `positron_`
- I'm imagining the `register_method()` function will be common to both ark and positron features, so should be namespaced as such.

We probably should meet and discuss how to expose this API.

dfalbel commented on crates/ark/src/interface.rs:
```diff
@@ -403,6 +405,9 @@ impl RMain {
             }
 
+            populate_methods_from_loaded_namespaces()
+                .or_log_error("Can't populate variables pane methods from loaded packages");
```
Resolved in https://github.com/posit-dev/ark/pull/560/commits/8551bf3e102fce54d335ca1851cb683e68c4ae39

dfalbel commented on crates/ark/src/variables/variable.rs:
```diff
@@ -1266,6 +1328,16 @@ impl PositronVariable {
+    let kind: Option<String> = ArkGenerics::VariableKind.try_dispatch(value, vec![])?;
+    match kind {
+        None => Ok(None),
+        // Enum reflection is not beautiful, we want to parse a VariableKind from it's
```
This was more like a rant :P.  It's hard to go from the string representation to an enum value in rust. Here we had to create a json and then read it with serde.

dfalbel commented on crates/ark/src/variables/methods.rs:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    pub fn register_method(generic: Self, class: &str, method: RObject) -> anyhow::Result<()> {
```
My reasoning was that a static method would make more clear that we're actually modfying some global state and not just an object state. But loooks weird indeed - fixed in https://github.com/posit-dev/ark/pull/560/commits/16d4ab891c7df524e1f3ebf1f875b2cba2c48564

dfalbel commented on crates/ark/src/variables/methods.rs:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    pub fn register_method_from_package(
+        generic: Self,
```
fixed in https://github.com/posit-dev/ark/commit/16d4ab891c7df524e1f3ebf1f875b2cba2c48564

t-kalinowski commented on crates/ark/src/variables/methods.rs:
```diff
@@ -0,0 +1,158 @@
+                r_symbol!(":::"),
+                r_symbol!(package),
+                r_symbol!(format!("{generic}.{class}")),
+            )
```
IIRC, the motivation for storing a call `pkgname:::methodname` instead of the closure object was to ensure that:

1.  Everything worked seamlessly with `devtools::load_all()` and similar functions (no risk of calling an outdated method from a stale cache);
2.  An escape hatch remained available for monkey-patching a method in the package namespace (e.g., to enable workarounds for misbehaving methods).

dfalbel replied:
I updated to use `RCall`  in https://github.com/posit-dev/ark/pull/560/commits/aadef2b97c1f0c16fb528a5c13a0ac90e6b529a7

dfalbel commented on crates/ark/src/variables/methods.rs:
```diff
@@ -0,0 +1,158 @@
+    pub fn try_dispatch<T>(
+        &self,
+        x: SEXP,
+        args: Vec<(String, RObject)>,
```
Done in https://github.com/posit-dev/ark/pull/560/commits/51111c00453242d50b5c47a191805c88300fe4ab

dfalbel commented on crates/ark/src/variables/methods.rs:
```diff
@@ -0,0 +1,158 @@
+use crate::modules::ARK_ENVS;
+
+#[derive(Debug, PartialEq, EnumString, EnumIter, IntoStaticStr, Display, Eq, Hash, Clone)]
+pub enum ArkGenerics {
```
Do you have a preference? It feels like this could be used elsewhere, so we could move to its own module above.

lionel- replied:
I'm happy with this change, it seems like that infrastructure could be useful in the future.

dfalbel commented on crates/ark/src/modules/positron/methods.R:
```diff
@@ -0,0 +1,56 @@
+
+    for (cl in class(object)) {
+        if (!is.null(method <- get0(cl, envir = methods_table))) {
+            return(eval(
+                as.call(list(method, object, ...)),
+                envir = globalenv()
+            ))
```
As it's currently implemented `method` is not necessarily a function. It could be a call object, a symbol or a function.

lionel- replied:
A symbol? To be evaluated in global? hmmm

I think we should evaluate the input in a child of base then, and bind to a mask to avoid the inlining.

Are you sure that symbolic objects are not evaluated on the Rust side at the `call_in()` site? Are they protected from evaluation?

dfalbel replied:
In theory, a user could do 

```
.ark.register_method("positron_display_values", "foo", quote(my_method_name)))
```

and that would call `my_method_name(x)` in the global env when dispatched.
I'm not sure I fully understand the risks of this approach.

dfalbel commented:
Thanks @lionel- 

I agree that we should have the opposite namespacing and just changed here: https://github.com/posit-dev/ark/pull/560/commits/a21dc3748cad86b1c9c589161264f16fa3962b7f

I'm happy to chat further on how to expose this API. Currently we don't necessarily need to expose `.ark_register_method()` as we already automatically find methods in packages, but it's nice for interactive debugging and for testing purposes.

DavisVaughan commented on crates/ark/src/methods.rs:
```diff
@@ -0,0 +1,154 @@
+        })
+        .map(|b| -> String { b.name.into() });
+
+    for name in symbol_names {
+        if let Some((generic, class)) = ArkGenerics::parse_method(&name) {
```
It feels kind of inefficient to me to have to generate the vector of `String`s for each `ArkGeneric` each time you go through `parse_method()`. It's like we want to generate the vector of generic names once up front and reuse them for all symbol names we compare against.

dfalbel replied:
Removed automatic registration in https://github.com/posit-dev/ark/pull/560/commits/5ab5b9dcd0bca352bd0db6c369bf8eae33896d88

DavisVaughan reviewed (commented):
Partial review - will do more in a bit

t-kalinowski commented on crates/ark/src/modules/positron/methods.R:
```diff
@@ -0,0 +1,55 @@
+.ark.register_ark_method <- function(generic, class, method) {
+    stopifnot(
+        is_string(generic),
+        generic %in% c(
+            "positron_variable_display_value",
+            "positron_variable_display_type",
+            "positron_variable_has_children",
+            "positron_variable_kind"
+        ),
```
During interactive development, it’s convenient to quickly peek the list of method names by simply printing `.ark.register_ark_method`. Maybe we could still accommodate that by including the list of method names in an error message.

DavisVaughan replied:
Yea I don't think we should let that stop us from trying to remove this source of potentially-error-prone duplication

DavisVaughan commented on crates/ark/src/variables/variable.rs:
```diff
@@ -312,6 +320,34 @@ impl WorkspaceVariableDisplayValue {
+            )]);
+
+        let display_value = unwrap!(display_value, Err(err) => {
+            log::error!("Failed to apply 'ark_variable_display_value': {err:?}");
```
```suggestion
            log::error!("Failed to apply 'positron_variable_display_value': {err:?}");
```

or `ark_positron_`, whatever we decide

dfalbel replied:
Fixed in https://github.com/posit-dev/ark/pull/560/commits/fdee0f42f4727615f5f3b31a4aa41c061f849781

DavisVaughan commented on crates/ark/src/variables/variable.rs:
```diff
@@ -327,6 +363,12 @@ impl WorkspaceVariableDisplayType {
     ///   display type.
     pub fn from(value: SEXP, include_length: bool) -> Self {
+        match Self::try_from_method(value, include_length) {
+            Err(e) => log::error!("Error from 'ark_variable_display_type' method: {e}"),
```
`positron_` or whatever we decide on

dfalbel replied:
Fixed in https://github.com/posit-dev/ark/pull/560/commits/fdee0f42f4727615f5f3b31a4aa41c061f849781

DavisVaughan commented on crates/ark/src/variables/variable.rs:
```diff
@@ -434,6 +485,12 @@ impl WorkspaceVariableDisplayType {
 
 fn has_children(value: SEXP) -> bool {
+    match ArkGenerics::VariableHasChildren.try_dispatch(value, vec![]) {
+        Err(e) => log::error!("Error from 'ark_variable_has_children' method: {e}"),
```
`positron_` or whatever we decide on

dfalbel replied:
Fixed in https://github.com/posit-dev/ark/pull/560/commits/fdee0f42f4727615f5f3b31a4aa41c061f849781

DavisVaughan commented on crates/ark/src/variables/variable.rs:
```diff
@@ -625,6 +682,12 @@ impl PositronVariable {
         }
 
+        match try_from_method_variable_kind(x) {
+            Err(e) => log::error!("Error from 'ark_variable_kind' method: {e}"),
```
`positron_` or whatever we decide on

dfalbel replied:
I've made this change in https://github.com/posit-dev/ark/pull/560/commits/fdee0f42f4727615f5f3b31a4aa41c061f849781

DavisVaughan commented on crates/ark/src/variables/variable.rs:
```diff
@@ -1266,6 +1329,16 @@ impl PositronVariable {
+    let kind: Option<String> = ArkGenerics::VariableKind.try_dispatch(value, vec![])?;
+    match kind {
+        None => Ok(None),
+        // We want to parse a VariableKind from it's string representation.
```
```suggestion
        // We want to parse a VariableKind from its string representation.
```

dfalbel commented on crates/ark/src/methods.rs:
```diff
@@ -0,0 +1,154 @@
+#[derive(Debug, PartialEq, EnumString, EnumIter, IntoStaticStr, Display, Eq, Hash, Clone)]
+pub enum ArkGenerics {
+    #[strum(serialize = "positron_variable_display_value")]
+    VariableDisplayValue,
```
Made this change in https://github.com/posit-dev/ark/pull/560/commits/fdee0f42f4727615f5f3b31a4aa41c061f849781

dfalbel commented on crates/ark/src/methods.rs:
```diff
@@ -0,0 +1,154 @@
+
+    for pkg in loaded.into_iter() {
+        populate_methods_table(pkg.as_str()).or_log_error("Failed populating methods");
+    }
```
Removed auto registration in https://github.com/posit-dev/ark/pull/560/commits/5ab5b9dcd0bca352bd0db6c369bf8eae33896d88

dfalbel commented:
As discussed, I have updated the PR with:

- Removing auto-registration in favor of only explicit method registration.
- Added an allowed list of packages, so we can test the API for some time before making it really public.

I have torch PR implementing how packages should explicitly register methods: https://github.com/mlverse/torch/pull/1200/commits/faa11e4a1364884a11afee616e01745c48923e5c

t-kalinowski commented:
@lionel- @DavisVaughan Can I ask why
- automatic discovery of methods in packages was removed,
- method re-registration via `devtools::load_all()` was (intentionally?) removed,
- and only packages on the explicit allow list are allowed to register methods

I’d like to implement methods not only in reticulate but also in keras and tensorflow. The recent changes before merge feel like unnecessary added complexity, inconvenience, and communication overhead, for reasons I don’t fully understand.

lionel- commented:
(1) is a choice for explicitness. This API seems sufficiently specific that magic and extra convenience did not feel warranted.
(2) is about not committing to any public interface in the short term until the mlverse had time to experiment with the API.

t-kalinowski commented:
That makes sense, thanks.

wesm commented:
I was reading through this since it was mentioned in https://github.com/posit-dev/positron/issues/5573. 

A side comment: I wanted to mention that in https://github.com/posit-dev/ark/blob/main/doc/variables-pane-extending.md, Positron is not mentioned at all. Ark contains a lot of code that powers features in Positron but not easy to use elsewhere. Since Ark is intended to be usable outside of Positron, it may make sense to more clearly separate out the Positron-specific code and documentation so that people don't get the wrong idea that Ark is only relevant to Positron.
